src/RevLoader.cc

```diff
@@ -143,24 +143,46 @@ bool RevLoader::LoadElf32(char *membuf, size_t sz){
 
   // Add memory segments for each program header
   for (unsigned i = 0; i < eh->e_phnum; i++) {
-    // Print information about the program header
-    // std::cout << "Program Header " << i << ": " << ph[i].p_type << ", Address: " << ph[i].p_paddr << ", Size: "  << std::endl;
-
-    // Add a memory segment for the program header
     if( sz < ph[i].p_offset + ph[i].p_filesz ){
       output->fatal(CALL_INFO, -1, "Error: RV32 Elf is unrecognizable\n" );
     }
-    uint32_t SegSize = (ph[i].p_memsz < __PAGE_SIZE__) ? __PAGE_SIZE__ : ph[i].p_memsz;
-    mem->AddMemSeg(ph[i].p_paddr, SegSize, true);
+    // Add a memory segment for the program header
+    mem->AddMemSeg(ph[i].p_paddr, ph[i].p_filesz, true);
   }
 
-  // Add memory segments for each section header
+  uint64_t StaticDataEnd = 0; 
+  uint64_t BSSEnd = 0; 
+  uint64_t DataEnd = 0; 
+  uint64_t TextEnd = 0; 
+  // Add memory segments for each section header 
+  // - This should automatically handle overlap and not add segments
+  //   that are already there from program headers
   for (unsigned i = 0; i < eh->e_shnum; i++) {
-    // Print information about the program header
-    // std::cout << "Section Header " << i << ": " << sh[i].sh_type << ", Address: " << sh[i].sh_addr << ", Size: "  << std::endl;
+    // check if the section header name is bss
+    if( strcmp(shstrtab + sh[i].sh_name, ".bss") == 0 ){
+      BSSEnd = sh[i].sh_addr + sh[i].sh_size;
+    }
+    if( strcmp(shstrtab + sh[i].sh_name, ".text") == 0 ){
+      TextEnd = sh[i].sh_addr + sh[i].sh_size;
+    }
+    if( strcmp(shstrtab + sh[i].sh_name, ".data") == 0 ){
+      DataEnd = sh[i].sh_addr + sh[i].sh_size;
+    }
     mem->AddMemSeg(sh[i].sh_addr, sh[i].sh_size, true);
   }
-
+  // If BSS exists, static data ends after it
+  if( BSSEnd > 0 ){
+    StaticDataEnd = BSSEnd;
+  } else if ( DataEnd > 0 ){
+    // BSS Doesn't exist, but data does
+    StaticDataEnd = DataEnd;
+  } else if ( TextEnd > 0 ){
+    // Text is last resort 
+    StaticDataEnd = TextEnd;
+  } else {
+    // Can't find any (Text, BSS, or Data) sections
+    output->fatal(CALL_INFO, -1, "Error: No text, data, or bss sections --- RV64 Elf is unrecognizable\n" );
```

Comment:
Noob question - are these sections always exclusive?

Reply:
Yes. The same sections may be present within the same segment however as I understand it:
- `.text`: Contains the actual executable code (ie. Always present)
- `.data`: Contains initialized global and static variables 
- `.bss`: Contains uninitialized global and static variables

The latter two are absent from the `isa` tests, however, should be present in most other programs.